Add expanded description of versioning (#126)
memfault/memfault-docs: docs/platform/software-version-hardware-version.mdx

  firmware, the firmware's Software Version is used to determine if a device
  requires an update

### Version Schemes

> 

Author: todo add supported character list:
`-a-zA-Z0-9_.+: ()[]/`

for more information).

For example, these versions are sorted from oldest to newest:

```plaintext
1.0.0rc1
1.0.0
1.0.1-dev
1.0.1-rc2
1.0.1
1.0.12
1.0.20
```

To implement a "release candidate" or "alpha/beta" versioning scheme, the
recommended option is to follow the semver 2.0 specification, for example:

```plaintext
1.0.0-alpha.1
1.0.0-alpha.2
```

Or:

```plaintext
1.0.0-rc.0
1.0.0-rc.1
```

Be cautious about more exotic versioning schemes! natsort will only handle the
common scenarios.

For example, the semver 2.0 spec permits appending build metadata to a version
number:

```plaintext
1.0.0+092ced197518c8e304232239315c611af6ff2404
```

But natsort will apply lexical ordering (under the default settings), which may
not result in the expected OTA response (see _"Disable Version Checking"_ above
for a workaround).

#### Recommended Versioning Scheme

Memfault recommends using [Semver 2.0](https://semver.org/) for versioning.
Specifically, using these forms will result in the least confusion:

1. for official releases, strict semver ("normal version number" per the spec):

   ```plaintext

>    1.2.3

Reviewer: ```suggestion
   1.0.0
```

>    2022.04.02

Reviewer: might drop this one and just leave the 1.2.3

   22.10.0
   ```

   Exactly 3 digit groups, leading zeros permitted, delimited by periods (see
   the spec for details).

2. for development releases, depending on your needs:

   1. rapid iteration that requires downgrade protection and filtering/grouping
      of devices based on version:

      ```plaintext
      1.2.3-rc.0
      1.2.4-rc.2
      ```

   2. rapid iteration that doesn't require downgrade protection:

      ```plaintext

>       1.2.3+092ced197518c8e304232239315c611af6ff2404

Reviewer: For development builds, I've become a fan of the .99 strategy

i.e pre 1.1 release:

```suggestion
      1.0.99
```

or aleternatively

```
1.1.0-dev
```

Author: I was thinking about the case where they're pushing out eg. nightlies to a cohort- the version identifier would need to change in that case I think? could use incrementing `1.2.3-dev.0` etc, but in case they didn't have CI set up for that... dunno what the best strategy would be :(

Author: i expanded this section with a few other strategies... could be split off into its own doc, dunno what makes sense!

Author: also filled it with some fairly opinionated stuff so feel free to push any changes :sweat_smile:

      ```

   And set the _"Disable Version Checking"_ setting for the cohort.

#### Testing

To test version ordering, two approaches are recommended:

1. locally, using the `natsort`
   [command line utility](https://natsort.readthedocs.io/en/master/shell.html):

   ```bash
   ❯ natsort 2022.04.01 2022.10.02 2022.05.05 2022.05.04
   2022.04.01
   2022.05.04
   2022.05.05
   2022.10.02
   ```

>    \_Note that mimic the way Memfault's system operates, it may be necessary to

Reviewer: ```
from natsort import natsorted
natsorted(["1.0.1", "1.0.2"], key=lambda x: x + "z")
['1.0.1', '1.0.2']
```

Reviewer: ```suggestion
   \_Note For exotic versioning schemes, to mimic the way Memfault's system operates, it may be necessary to
```

  firmware, the firmware's Software Version is used to determine if a device
  requires an update

### Version Schemes

#### Ordering

> We recommend using a [SemVer](https://semver.org/) for your software version.
> (See section below)

Reviewer 2: Super nit: actually link to the section, or otherwise improve punctuation

```suggestion
We recommend using a [SemVer](https://semver.org/) for your software version (see section below).
```

If you do not use SemVer, we fallback to using natural sort, as implemented by
[natsort](https://pypi.org/project/natsort/) to resolve ordering.

This means that a device will not receive a previous OTA release, even if it's
deployed to the cohort (or the device cohort changes); in other words,
**downgrades are not permitted**.

To override this logic, there's a per-cohort setting _"Disable Version
Checking"_ which will always return the currently-deployed release to a device,
even if it's older (downgrades will then be possible! so be sure the affected
devices can tolerate the downgrade).

Memfault uses a slight tweak on top of `natsort` to support `rc` or `dev`

> versioning schemes. This is achieved by appending a character `z` (see
> [here](https://natsort.readthedocs.io/en/master/examples.html#sorting-more-expressive-versioning-schemes)
> for more information).

Reviewer 2: Super nit:
```suggestion
versioning schemes. This is achieved by [appending a character `z`](https://natsort.readthedocs.io/en/master/examples.html#sorting-more-expressive-versioning-schemes).
```
A separate sentence for the link isn't necessary IMO.

For example, these versions are sorted from oldest to newest:

```plaintext
1.0.0rc1
1.0.0
1.0.1-dev
1.0.1-rc2
1.0.1
1.0.12
1.0.20
```

To implement a "release candidate" or "alpha/beta" versioning scheme, the
recommended option is to follow the semver 2.0 specification, for example:

```plaintext
1.0.0-alpha.1
1.0.0-alpha.2
```

Or:

```plaintext
1.0.0-rc.0
1.0.0-rc.1
```

Be cautious about more exotic versioning schemes! natsort will only handle the
common scenarios.

For example, the semver 2.0 spec permits appending build metadata to a version
number:

```plaintext
1.0.0+092ced197518c8e304232239315c611af6ff2404
```

But natsort will apply lexical ordering (under the default settings), which may
not result in the expected OTA response (see _"Disable Version Checking"_ above

> for a workaround).

Reviewer 2: Nit: link instead? Less prose, and easier to navigate.

   ```bash
   # using a build number (eg CI build number, or monontonically incrementing
   # based on git tag, etc.)
   1.0.0-nightly.0

   # using a calendar version. always increments, and can be generated without
   # context (assuming only 1 build is deployed per day)
   1.0.0-nightly.20220123

   ```

4. development builds generated locally (using git sha or other identifier):

   ```plaintext
   1.2.3+092ced197518c8e304232239315c611af6ff2404
   ```

   And set the _"Disable Version Checking"_ setting for the cohort.

:::info

Note that for these schemes, the `device_info.software_version` field may need
to be populated at build time by the build script. For example, you might have a
C preprocessor token that holds the string for the software version, where the
value is populated by a build script:

```bash
# a C compiler option setting the SOFTWARE_VERSION token to a string
-DSOFTWARE_VERSION=\"1.0.0-nightly-20220123\"`
```

Some potentially useful things to consider:

- if using git tags, see information on `git rev-parse` for generating tag
  strings
- if using a calendar version, the unix `date` command could be used, for

>   example `date --utc +%Y%m%d`

Reviewer 2: Super nit: missing period on the final bullet point.
```suggestion
  example `date --utc +%Y%m%d`.
```

of the Firmware SDK for details on how to set this up on the device's software
side.

#### Android

Memfault supports extracting the Software Version for the `android-build`
Software Type from several different sources in order to suit different build
configurations.

> By default, irrespective of the source, the Software Version will be ordered per
> the [Version Schemes](#version-schemes) section above.

Reviewer 2: Super nit: the link makes this redundant, and we may move the section.

```suggestion
By default, irrespective of the source, the Software Version will be ordered per
[Version Schemes](#version-schemes).
```
